docs/grammar.rst

```diff
@@ -0,0 +1,256 @@
+Grammar syntax
+==============
+
+The grammar consists of a sequence of rules of the form:
+
+::
+
+       rule_name: expression
+
+Optionally, a type can be included right after the rule name, which
+specifies the return type of the Python function corresponding to the
+rule:
+
+::
+
+       rule_name[return_type]: expression
+
+If the return type is omitted, then ``Any`` is returned.
+
+Grammar Expressions
+-------------------
+
+``# comment``
+~~~~~~~~~~~~~
+
+Python-style comments.
+
+``e1 e2``
+~~~~~~~~~
+
+Match e1, then match e2.
+
+::
+
+       rule_name: first_rule second_rule
+
+.. _e1-e2-1:
+
+``e1 | e2``
+~~~~~~~~~~~
+
+Match e1 or e2.
+
+The first alternative can also appear on the line after the rule name
+for formatting purposes. In that case, a \| must be used before the
+first alternative, like so:
+
+::
+
+       rule_name[return_type]:
+           | first_alt
+           | second_alt
+
+``( e )``
+~~~~~~~~~
+
+Match e.
+
+::
+
+       rule_name: (e)
+
+A slightly more complex and useful example includes using the grouping
+operator together with the repeat operators:
+
+::
+
+       rule_name: (e1 e2)*
+
+``[ e ] or e?``
+~~~~~~~~~~~~~~~
+
+Optionally match e.
+
+::
+
+       rule_name: [e]
+
+A more useful example includes defining that a trailing comma is
+optional:
+
+::
+
+       rule_name: e (',' e)* [',']
+
+.. _e-1:
+
+``e*``
+~~~~~~
+
+Match zero or more occurrences of e.
+
+::
+
+       rule_name: (e1 e2)*
+
+.. _e-2:
+
+``e+``
+~~~~~~
+
+Match one or more occurrences of e.
+
+::
+
+       rule_name: (e1 e2)+
+
+``s.e+``
+~~~~~~~~
+
+Match one or more occurrences of e, separated by s. The generated parse
+tree does not include the separator. This is otherwise identical to
+``(e (s e)*)``.
+
+::
+
+       rule_name: ','.e+
+
+.. _e-3:
+
+``&e``
+~~~~~~
+
+Succeed if e can be parsed, without consuming any input.
+
```

Comment:
One thing missing here, and in much documentation of pegen that I've seen (PEP etc.), is what `&&` means. It's used a bunch in cpython and data's `python.gram`.

Hmm, but somehow it's documented in `grammar.md`. Should that redundant file be removed?

(I tried to look for other such undocumented symbols, but I don't see any.)

Reply:
> , is what `&&` means. It

Ah, excellent point. This was the only grammar element that was added **after** the PEP. I will make sure to add it there.

> Hmm, but somehow it's documented in `grammar.md`. Should that redundant file be removed?

Indeed, I will merge these two files.

Reply:
I pushed a change merging both files

Reply:
Great! Some important other additions here, like documenting `LOCATIONS`, `@class`, and `@subheader`.  I added a few specific comments to [the commit](https://github.com/we-like-parsers/pegen/commit/0328de304e19fd931726d213fbc88b5a2e1dab68).